apps/jupyter/extensions/notebook-runner/src/packages.ts

```diff
@@ -0,0 +1,208 @@
+      // ignore status messages
+      if (KernelMessage.isStatusMsg(msg)) {
+        return;
+      } else if (KernelMessage.isStreamMsg(msg)) {
+        console.log(`[${kernel.id}]`, msg.content.text);
+        return;
+      } else if (KernelMessage.isExecuteInputMsg(msg)) {
+        console.debug(
+          `[${kernel.id}] Executing code (${msg.content.execution_count}):\n`,
+          msg.content.code,
+        );
+        return;
+      } else if (KernelMessage.isExecuteResultMsg(msg)) {
+        console.debug(
+          `[${kernel.id}] Finished executing code (${msg.content.execution_count}):`,
+          msg.content.data,
+        );
+      }
+
+      console.debug(`[${kernel.id}] IOPub message:`, msg);
+    });
+
+    console.debug("Installing packages...", kernel.id);
+    await importBasePackages(kernel);
+    await installOtter(kernel);
+    console.debug("Finished installing packages", kernel.id);
+
+    if (
+      notebookPath === EmbeddedPythonCallbacks.studentTaskLocation ||
+      notebookPath === EmbeddedPythonCallbacks.studentTaskLocation.slice(1)
+    ) {
+      console.debug(
+        "Opened student notebook - copying notebook to the virtual filesystem to make tests available..",
+      );
+
+      let notebook: Contents.IModel | null = null;
+      try {
+        notebook = await contentsManager.get(notebookPath, { content: true });
+      } catch (error) {
+        throw new Error("Error reading ran notebook:" + error);
+      }
+
+      await writeJsonToVirtualFilesystem(
+        kernel,
+        EmbeddedPythonCallbacks.studentTaskLocation,
+        notebook.content,
+      );
+      console.debug("Finished copying notebook to the virtual filesystem");
+
+      await kernel.requestExecute({
+        code: `os.chdir("/student")`,
+      }).done;
+    }
+  };
+
+const trackSession = (
+  contentsManager: ContentsManager,
+  sessionContext: ISessionContext,
+  notebookPath: string,
+): Promise<void> =>
+  addKernelListeners(
+    sessionContext,
+    autoInstallPackages(contentsManager, notebookPath),
+  );
+
+export const preInstallPackages = async (
+  _app: JupyterFrontEnd,
+  contentsManager: ContentsManager,
+  notebookTracker: INotebookTracker,
+): Promise<void> => {
+  notebookTracker.forEach((panel) =>
+    trackSession(contentsManager, panel.sessionContext, panel.context.path),
+  );
+
+  notebookTracker.widgetAdded.connect((_, panel) => {
+    // get path of notebook
+    const notebookPath = panel.context.path;
+
+    trackSession(contentsManager, panel.sessionContext, notebookPath);
+  });
+};
+
+export const importBasePackages = async (
+  kernel: IKernelConnection,
+): Promise<void> => {
+  console.log("Importing builtin libraries..");
+  await kernel.requestExecute(
+    {
+      code: `
+        import micropip, sys, types, os, tempfile, asyncio, shutil, base64
+        from urllib.parse import unquote
+        from pathlib import Path
+      `,
+    },
+    true,
+  ).done;
+};
```

Comment:
I don't understand why this is necessary. Is this a default for Jupyter?
Some documentation concerning the intent might be helpful here.